### PR TITLE
Setup workflow for building & publishing docker image

### PR DESCRIPTION
**Pull request by alec-pinson**
- https://github.com/alec-pinson/TerrariumPI/issues/22

### CODE REVIEW

**Comment by alec-pinson** [comment]
@theyosh this will automatically build the docker image and push it to dockerhub.

There's a couple of pre-reqs you'll need to do first
1. Setup a Dockerhub account
2. Create an access token
3. Add your username and the access token to your GitHub repository secrets named `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN`
4. update the tags line on `.github/workflows/docker.yaml` https://github.com/alec-pinson/TerrariumPI/blob/ed84cddf0d5f6d0bae51ecad70db815733a72448/.github/workflows/docker.yaml#L37

Looking at your branch setup, `master` should really always be the `latest` version and then branches should be for new features.
I understand you still have the old version 3 and so maybe keeping a branch for this is ok.

I see you have a `documentation` branch and was wondering how come this isn't part of `master`?
For the documentation workflow you could have something like
```yaml
on:
  push:
    branches:
      - master
    paths:
      - 'docs/**'
```
and for the other workflows
```yaml
on:
  push:
    branches:
      - master
    paths-ignore:
      - 'docs/**'
```

Another side note, GitHub changed their default branch naming from `master` to `main` quite a while ago. Not sure if you'd want to follow suit and rename `master`.

Anyway the reason I bring up your branching strategy is because when pushing the docker image you need to decide how to automate the docker tag.

First thing to say about docker tags is that you can have multiple tags for an image and docker is clever enough to not require you to build/push the whole image for each tag. So for example you could have your tagging setup like this if using feature branches:-

1. Create branch called `magical-new-sensor`
2. Add magical new sensor code/libs
3. Push/create PR, docker image can automatically be created and published with
- theyosh/terrariumpi:magical-new-sensor
- theyosh/terrariumpi:development
- theyosh/terrariumpi:<COMMIT-SHA>
4. Normally you would deploy/run the container somewhere and some testing would be completed here (more complicated with Raspberry Pis, see below)
5. You decide you're happy and merge the PR
6. You can decide now if you want it to automatically build/publish once on master or another option is automatically build/publish once you've created your release which would mean automatic tagging like:
- theyosh/terrariumpi:v4.0.1
- theyosh/terrariumpi:<COMMIT-SHA>
- theyosh/terrariumpi:latest

### Testing
There is an option to try to setup some testing as part of a workflow if you wanted to.. GitHub support something called `self-hosted runners`. Basically you would have a dedicated raspberry pi for testing and set it up as a self-hosted runner on your repository. Once the docker image is built you can automatically deploy it to your raspberry pi and confirm everything works, if it doesn't you can fail the build.


Another thing I was wondering is if you've ever used discord?/maybe you should setup a TP discord channel. It would make some things alot easier being able to instant chat 😃

**Comment by theyosh** [comment]
> Another thing I was wondering is if you've ever used discord?/maybe you should setup a TP discord channel. It would make some things alot easier being able to instant chat smiley

I have a Matrix server with a TP room :) https://matrix.to/#/#terrariumpi:theyosh.nl using Element as client: https://app.element.io/

**Comment by theyosh** [comment]
About the branches, I used documentation for generating the website: https://theyosh.github.io/TerrariumPI/ which is my documentation / manual info site. I put this in a different branch, because else the Github action for building the documentation will be run multiple times during a day, even if there are no updates. That would waste a lot of CPU time. Also, it would change the build date on the documentation pages, so it looks like it is updating all the time, which is not true.

So, that is why I have a documentation branch. But that should be a 'mirror' of the 4.x.y.z branch. https://github.com/theyosh/TerrariumPI/actions/workflows/pages-deploy.yml

The master branch is still a 'mirror' of the 3.x.y.z branch. This is still the case, but next month, when I hope to release 4.1 (with also Docker), I will merge 4.x.y.z to master. It waited so long, so that people can prepare the upgrade. And there is not a real upgrade path from 3 to 4. So if somebody is just updating the master branch, they will get into issues. So that is why I wait a few months, and at least 1 release, so that everybody can either switch to the 3.x.y.z branch, and stay there. Or do a reinstall with version 4. But the master branch should become the active and 4.x version.

And I was thinking of making only docker images per release. As I push a lot of small fixes to the master branch. And that could case some bugs. And when other people notice that, I can fix that pretty fast. But making a docker image all the time is not smart. So I would use docker for only releasing releases, and not master or PR versions.

So, I would trigger the Docker build on tags I think.

**Comment by alec-pinson** [comment]
> > Another thing I was wondering is if you've ever used discord?/maybe you should setup a TP discord channel. It would make some things alot easier being able to instant chat smiley
> 
> I have a Matrix server with a TP room :) https://matrix.to/#/#terrariumpi:theyosh.nl using Element as client: https://app.element.io/

Oh I didnt see this lol

> About the branches, I used documentation for generating the website: https://theyosh.github.io/TerrariumPI/ which is my documentation / manual info site. I put this in a different branch, because else the Github action for building the documentation will be run multiple times during a day, even if there are no updates. That would waste a lot of CPU time. Also, it would change the build date on the documentation pages, so it looks like it is updating all the time, which is not true.

if you add `paths-ignore` to your workflows then it shouldn't run whenever docs is updated.
```yaml
on:
  push:
    branches:
      - master
    paths-ignore:
      - 'docs/**'
```

> The master branch is still a 'mirror' of the 3.x.y.z branch. This is still the case, but next month, when I hope to release 4.1 (with also Docker), I will merge 4.x.y.z to master. It waited so long, so that people can prepare the upgrade. And there is not a real upgrade path from 3 to 4. So if somebody is just updating the master branch, they will get into issues. So that is why I wait a few months, and at least 1 release, so that everybody can either switch to the 3.x.y.z branch, and stay there. Or do a reinstall with version 4. But the master branch should become the active and 4.x version.

Ah ok fair enough.

> And I was thinking of making only docker images per release. As I push a lot of small fixes to the master branch. And that could case some bugs. And when other people notice that, I can fix that pretty fast. But making a docker image all the time is not smart. So I would use docker for only releasing releases, and not master or PR versions.
> 
> So, I would trigger the Docker build on tags I think.

You definitely need to Docker build on PR to confirm the Docker build is successful prior to merge, you don't have to publish it to dockerhub though if you don't want to.

**Comment by theyosh** [comment]
> You definitely need to Docker build on PR to confirm the Docker build is successful prior to merge, you don't have to publish it to dockerhub though if you don't want to.

Not sure how to do this. And I have a problem with this, as it takes a lot of time, and I can only test if it works with a physical Pi with hardware connected. So I do not see much value in testing a docker image just to see if it starts.... as that is not changing.

And most PR is about new hardware, which is 100% python code. And that I test on my pi, and if that works, it is good enough for me. Building docker image should just work, as it only has to update python code.

And, it is a one man product. And I do not want to spent more time on testing and building then developing. I am still a developer. :) So not sure how to do all this on my own..

File: .github/workflows/docker.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        name: Set up Docker Buildx
+        if: steps.Dockerfile.outputs.files_exists == 'true'
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Docker meta
+        if: steps.Dockerfile.outputs.files_exists == 'true'
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: wkd1/test
```
Comment (theyosh) [inline .github/workflows/docker.yaml:46]: This is the only change when merged?

**Comment by alec-pinson** [comment]
> > You definitely need to Docker build on PR to confirm the Docker build is successful prior to merge, you don't have to publish it to dockerhub though if you don't want to.
> 
> Not sure how to do this. And I have a problem with this, as it takes a lot of time, and I can only test if it works with a physical Pi with hardware connected. So I do not see much value in testing a docker image just to see if it starts.... as that is not changing.
> 
> And most PR is about new hardware, which is 100% python code. And that I test on my pi, and if that works, it is good enough for me. Building docker image should just work, as it only has to update python code.
> 
> And, it is a one man product. And I do not want to spent more time on testing and building then developing. I am still a developer. :) So not sure how to do all this on my own..

Haha ok fair enough 👍

File: .github/workflows/docker.yaml (hunk 1 above)
Comment (alec-pinson) [reply .github/workflows/docker.yaml]: yep, ive updated it for you and also changed from running on PR so it only runs on master/release creation

**Comment by alec-pinson** [comment]
By the way the reason I say build during PR is because if you merge and then the build fails you need to create a new PR to try fix, and if it fails again, rinse repeat.

**Pull request merged**